colorschemes/everforest: init

Sheemap opened the pull request:
Adding the everforest colorscheme https://github.com/sainnhe/everforest

Fairly new to nix, I would super appreciate any feedback yall have!

I pulled all the settings info strings directly out of `:help everforest`

GaetanLepage reviewed (changes requested):
This looks very solid to me at a first glance.
Could you please add a test ? (Refer to similar PRs/files to draw inspiration from).

GaetanLepage commented:
Looks like there is something wrong with the tests.
You can debug locally by entering the devshell (`nix develop` or using direnv).
Then, simply run the `tests` command.

Sheemap commented:
Got it, thank you!

So it looks like the issue is due to the `better_performance` flag being set to 1. This causes the everforest plugin to write stuff to the disk for caching or something.

Problem is Nix's immutability is not allowing that, and it fails. Theres no setting I can find to tweak the cache location.

Seems to me an upstream problem. What would you recommend I do to resolve this? Is there way to resolve this within Nixvim? Or perhaps just modifying the docs and raising an assertion error if that setting is set?

MattSturgeon commented:
> Problem is Nix's immutability is not allowing that, and it fails. Theres no setting I can find to tweak the cache location.

If so, that should only be affecting the test and should be fine when running normally? Unless the plugin is trying to write its cache to the nix store.

If it is just the test that's affected, you can disable whatever option is needed. Along with a comment, explaining why said option causes tests to fail.

Sheemap commented:
Unfortunately no its trying to write to the nix store. Can turn it off in the test, but this feature will not work in a real environment

```
line 5: Vim(call):E739: Cannot create directory /nix/store/0y403ixxd7jkdxvsjwl6alkayf7qxsza-vimplugin-everforest-2024-06-02/after: read-only file system                                                                                                                                                                                                                                         
```

MattSturgeon commented:
> Unfortunately no, its trying to write to the nix store. Can turn it off in the test, but this feature will not work in a real environment

Frustrating. We'll just have to disable the test and report an issue upstream, recommending they use `XDG_CACHE_HOME`. If an issue is open before this PR is done, we can reference it in a comment.

As for the option, we don't actually have to explicitly declare all availabile settings options anyway, since `settings` is a freeform option, users can define any config they like, even stuff we haven't added options for.

In some respects it's better not to define all available options; once we've added them we then have to try and keep them up to date.

Sheemap commented:
Cool sounds good! Ive removed the option from the settings. I'm happy to remove any others, though I'm not sure which ones are important enough to be kept in vs removed

MattSturgeon commented on tests/test-sources/plugins/colorschemes/everforest.nix:
```diff
@@ -0,0 +1,42 @@
+        inlay_hints_background = "dimmed";
+        disable_terminal_colors = 1;
+        lightline_disable_bold = 1;
+        colors_override = {
```
```suggestion
        # This option is broken, because the plugin tries to write its cache in /nix/store
        # better_performance = 1;
        colors_override = {
```

I'd rather have a comment rather than a silently missing option, but this is fine.

Just a preference not a request.

MattSturgeon replied:
If removing it is enough to fix the test, it makes me think that `1` _is not_ the default value anyway, making me second guess some of the other defaults listed here.

Sheemap replied:
Done! Thanks for the suggestion, definitely agree with your preference here.

MattSturgeon commented on plugins/colorschemes/everforest.nix:
```diff
@@ -0,0 +1,266 @@
+  ...
+}:
+let
+  mkEverforestBool =
+    pluginDefault: desc:
+    helpers.defaultNullOpts.mkEnum [
+      0
+      1
+    ] pluginDefault desc;
```
I think a type alias and/or option helper for `enum [0 1]` would be helpful in all of nixvim. Maybe in a future PR we'll do that

MattSturgeon commented on plugins/colorschemes/everforest.nix:
```diff
@@ -0,0 +1,266 @@
+
+      Currently, the following plugins are supported:
+
+      - neovim's built-in language server client
+      - https://github.com/neoclide/coc.nvim
+      - https://github.com/prabirshrestha/vim-lsp
+      - https://github.com/ycm-core/YouCompleteMe
+      - https://github.com/dense-analysis/ale
+      - https://github.com/neomake/neomake
+      - https://github.com/vim-syntastic/syntastic
```
I'd rather these be `[markdown links](url)`:

```suggestion
      - neovim's built-in language server client
      - [coc.nvim](https://github.com/neoclide/coc.nvim)
      - [vim-lsp](https://github.com/prabirshrestha/vim-lsp)
      - [YouCompleteMe](https://github.com/ycm-core/YouCompleteMe)
      - [ale](https://github.com/dense-analysis/ale)
      - [neomake](https://github.com/neomake/neomake)
      - [syntastic](https://github.com/vim-syntastic/syntastic)
```

Same for links in other option descriptions.

MattSturgeon commented on plugins/colorschemes/everforest.nix:
```diff
@@ -0,0 +1,266 @@
+
+    colors_override = helpers.defaultNullOpts.mkAttrsOf (lib.types.listOf lib.types.str) { } ''
+      Override color palette. The available keys can be found in
+      `https://github.com/sainnhe/everforest/blob/master/autoload/everforest.vim`
```
```suggestion
      The plugin's [source code](https://github.com/sainnhe/everforest/blob/master/autoload/everforest.vim).
```

MattSturgeon commented:
This is an excellent first time contribution! I hope you don't mine, I left some further feedback and discussion points.

- I'd prefer links to use markdown link syntax, so they are actually clickable in the docs and render nicely.
- I'd prefer a comment be left about the broken option (up to you)
- I'm confused as to why removing a default value has any effect; the whole point of the "plugin defaults" is they should match the plugin's defaults, as used when the option is not defined

Sheemap commented:
Thank you! :) Ive addressed all those concerns.

The `better_performance = 1` value is not the default, I actually intentionally made _all_ the test values non-default to potentially catch any that I didnt find on my mostly unmodified install (which worked in this case, wahoo :smile:) 

Do you expect the tests to be testing the default values? Is there a way I can create multiple tests in this file to test both the default and non-default cases?

Sheemap commented:
OH, I did not realize the test values are literally named as `plugin.defaults`! Only after re-reading your comment did I understand. Ill update those to be default values haha

MattSturgeon commented:
> Thank you! :) Ive addressed all those concerns.

Brilliant 

> Do you expect the tests to be testing the default values? Is there a way I can create multiple tests in this file to test both the default and non-default cases?

Yes you can create as many test as you like, generally "empty" and "defaults" are the bare minimum, but "example" is also commonly used to test non-defaults.

Tests can be named anything, those are just the most common names we use.

MattSturgeon reviewed (approved):
Great! Thanks for your persistence and patience!
